### ISSUE TITLE
Cannot use FileAsset in Dynamic provider

### ISSUE THREAD

**Issue by remeq** [pulumi/pulumi#6802]
## Summary

I am implementing my own resource and a dynamic provider which uses FileAsset as an input but pulumi fails on second run when it needs to deserialize the FileAsset object during the diff method and fails with error: 

 error: Exception calling application: Invalid asset encountered when unmarshalling resource property 

## Steps to reproduce

My resource is called AkClient accepts FileAsset as an Input and I call it like this. 

```
client = genesis_validator.AkClient('myclient',
    akt_script=pulumi.FileAsset('./test_script.sh'),
    conn = conn,
    vm_arn=arn)
```

In the create method I store all Inputs into the state with:

return CreateResult(id_='foo', outs={ 'cli_akt_addr': akt_addr, **inputs})

The first pulumi run passes,  and the file asset is stored in the state file: 

                       "akt_script": {
                            "4dabf18193072939515e22adb298388d": "c44067f5952c0a294b673a41bacd8c17",
                            "path": "./test_script.sh"
                        },

Second run fails with the error:

 error: Exception calling application: Invalid asset encountered when unmarshalling resource property 

## Context (Environment)

Python

## Affected feature

Dynamic provider, FileAsset

**Comment by leezen** [comment]
@remeq It'd be great if you could please share your use case around using `Asset` within your dynamic provider. Is a viable workaround to simply provide the path as a string and perform whatever file i/o operations you need within the provider itself?

**Comment by remeq** [comment]
I want my dynamic provider to perform actions only when the input file changes. I understood that this is what FileAsset is used for - that it computes hash of the file.

**Comment by leezen** [comment]
@remeq Yup. That makes sense -- thanks for clarifying what you're trying to do.

**Comment by remeq** [comment]
@leezen Ok :) so what is the conclusion? Is it a bug ?

**Comment by leezen** [comment]
Yes, I'd consider this a bug that we'll have to take a look at. In the meantime, I think the workaround would be to do as suggested, which would include computing the hash over the file as part of the dynamic provider.

**Comment by remeq** [comment]
Thanks, yes for now I implemented digest as a part of the dynamic provider.
